### ISSUE TITLE
rJava fails while invoking RJDBC library from OpenCPU API

### ISSUE THREAD

**Issue by itsamiths** [opencpu/opencpu#123]
**Goal :** To use RJDBC with OpenCPU AJAX Post call.  

**Scenario :** While invoking **RJDBC** library via OpenCPU Ajax Post call, **rJava** fails. 

**Environment :** public.opencpu.org

Here is what I am trying:

**html**

```
<textarea>
library(RJDBC)
.. do some data fetching and then r statistics
 </textarea>
<br />
<button id="submitbutton" >Execute</button>
<br />
<br />
<div id="output">
</div>
```

**javascript**

```
<script type="text/javascript">
  //When Document is Ready
    $(function () {
        var baseurl = "https://public.opencpu.org";
        //because identity is in base
        ocpu.seturl(baseurl+"/ocpu/library/base/R");
        //actual handler
        $("#submitbutton").on("click", function(){
            //arguments
            var mysnippet = new ocpu.Snippet($("#input").val());
            //perform the request
            var req = ocpu.call("identity", {
                "x" : mysnippet
            }, function(session){
                session.getStdout(function(outtxt){
                        $("#output-txt").text(outtxt);   
                });
            });   
            //if R returns an error, alert the error message
            req.fail(function(){
                alert("Server error: " + req.responseText);
                $("#output").hide();
            });      
            req.always(function(){
                $("button").removeAttr("disabled");  
            });
        });    
    });

</script>
```

**scenario**

when i click execute button, gets the user entered value from text area ( here library(RJDBC) and some db operations) and sends a ajax post and responds back with the below error

**google chrome console log :**

```
POST https://public.opencpu.org/ocpu/library/base/R/identity 400 (Bad Request) 
OpenCPU error HTTP 400
package 'rJava' could not be loaded
```

Hope this gives better clarity.

Need to confirm whether rJava works with OpenCPU, if so any possible work around to fix this?

StackOverflow : http://stackoverflow.com/questions/26999214/rjava-fails-while-invoking-rjdbc-library-from-opencpu-api
Google Groups : https://groups.google.com/forum/#!topic/opencpu/74Kth_ic3Mc

**Comment by ananthu2050** [comment]
I faced this same issue. I am building a opencpu App on top of mallet R package. mallet internally uses rJava. So when I try to call my opencpu App API, I get this same error "rjava could not be loaded". I cannot change mallet as I dont know its internals. So its a deadlock now. Any possible alternate solution available? Thanks

**Comment by greentheo** [comment]
...just adding in to the fray. Got RJDBC to work fine on windows for a webapp that connects to a  Vertica DB.  But yeah once I moved it onto the linux box rJava isn't working right and based on this thread I understand why.

Let me know if/how I can help solve the problem.

**Comment by GratienDSX** [comment]
Running into the same problem, using RJDBC to connect to a hive database. Any update on this?

In addition, I get this output when calling R code through ocpu api:

`
using 'as.environment(NULL)' is defunct

In call:
as.environment(pos)
`

**Comment by drknexus** [comment]
For those of you looking for a workaround, I've created a single node PSOCK cluster and have it start JDBC / issue commands through the cluster. It is cludgy, but it works.

**Comment by GratienDSX** [comment]
@drknexus thanks for your feedback. Not sure I understand the workaround however, can you give further details ?

**Comment by drknexus** [comment]
@GratienDSX The core of my solution looks like the below.  The code is taken from an R6 class that has some logging.  For the sake of documenting what is happening I've left in the `private$log` commands, but if running a version of this code you'd want to leave it out.  The code I provide here is not minimally replicable.  However the code I provide at [this gist](https://gist.github.com/drknexus/b5abf769527b5040817fbc946af2b2be) ought to be (once you provide the database connection, the `classPath` and `driverClass`).  That gist has been subject to about 0 clean-up, so there is a ton of code in there that is strictly speaking superfluous to getting the job done. 

Explanatory code segment:
```
jvmNode <- makeCluster(1, methods=FALSE, port = sample(11000:11999,1), rscript_args="--vanilla")
          connected <- isTRUE(try(clusterEvalQ(jvmNode, 2)==2, silent=TRUE))
        private$log$finest("forkable node started")
        clusterEvalQ(jvmNode,{library(DBI);library(rJava);library(RJDBC)})
        
        clusterEvalQ(jvmNode, systemRAMFree <- function() {
  #in GB, platform dependent 
  as.numeric(system('FREE_KB=$(($(echo `sed -n \'2p;3p;4p\' <  /proc/meminfo | sed "s/ \\+/ /g" | cut -d\' \' -f 2 ` | sed "s/ /+/g")));echo $FREE_KB', intern=TRUE))/1024/1024
})       
        clusterEvalQ(jvmNode, options(java.parameters = paste0("-Xmx",systemRAMFree(),"g")))

        hostInformation <- private$hostInformation
        classPath <- "~/redshift/redshiftJDBC41.jar"
        driverClass <- private$driverClass
        java.parameters <- getOption("java.parameters")
        toExportList <- list(statement,driverClass, classPath, hostInformation, java.parameters)
        names(toExportList) <- c("statement","driverClass","classPath", "hostInformation", "java.parameters")
        toExport <- as.environment(toExportList)
        clusterExport(jvmNode, ls(toExport), envir=toExport)
        clusterEvalQ(jvmNode, options(java.parameters = java.parameters))
        
        clusterEvalQ(jvmNode, rJava::.jinit())
        private$log$finest("JVM Started")
        clusterEvalQ(jvmNode, drv <- RJDBC::JDBC(driverClass, classPath))
        private$log$finest("Driver obtained")
        
        clusterEvalQ(jvmNode, conn <- dbConnect(drv, 
                                       url = hostInformation$host, 
                                       user = hostInformation$user, 
                                       password = hostInformation$password))
        private$log$debug(logPrefix("query:", statement))
        result <- try(clusterEvalQ(jvmNode, dbGetQuery(conn, statement))[[1]])
        clusterEvalQ(jvmNode, dbDisconnect(conn))
        stopCluster(jvmNode)
```

**Comment by GratienDSX** [comment]
@drknexus thanks a lot for sharing your solution, I'll benchmark quite soon against my other solution, that is very very nasty : instead of JDBC to access my Hive database I use ODBC and openquery to query Hive from R through SQL Server - at least it works.

**Comment by drknexus** [comment]
Oh, that's a clever approach.  I bet it could go faster your way provided the SQL Server shares a box with the R process.  Do let us know how it works out and (if possible) share the solution! :)

**Comment by ckanis** [comment]
I was struggling with this problem all day and wanted to post my solution for anyone who wants a quick and dirty fix because any Google search for my problem ended up at this page.  I needed to use RJDBC to make a select call to a database which only allowed a JDBC connection.  Even after making sure Apache knew where to look for java I was getting 'child process has died' errors.  My solution was to simply query the database using Rserve.

**Comment by jeroen** [comment]
I expect that these problems will be solved in OpenCPU 2.0 (which will be released in a week or so)

**Comment by jeroen** [comment]
Please open a new issue if this problem still exists in OpenCPU 2.0.
